src/ZoomableImage.android.js

```diff
@@ -0,0 +1,139 @@
+
+    this.startOffset = initialOffset;
+    this.startScale = initialScale;
+    this.startDistance = 0.1;
+
+    this.size = { width: 0, height: 0 };
+
+    this.panResponder = PanResponder.create({
+      onStartShouldSetPanResponder: (event, gestureState) => true,
+      onMoveShouldSetPanResponder: (event, gestureState) =>
+        gestureState.dx > 2 || gestureState.dy > 2 || gestureState.numberActiveTouches === 2,
+      onMoveShouldSetPanResponderCapture: (event, gestureState) => true,
+      onPanResponderGrant: (event, gestureState) => {
+        this.startOffset = this.offset;
+        this.startScale = this.scale;
+        if (gestureState.numberActiveTouches === 2) {
+          this.startDistance = calculateTouchDistance(event.nativeEvent.touches);
+        }
+      },
+      onPanResponderMove: (event, gestureState) => {
+        if (gestureState.numberActiveTouches === 1) {
+          const offset = this.calculateOffset(gestureState);
+          this.offset = offset;
+          this.offsetValue.setValue(offset);
+        } else if (gestureState.numberActiveTouches === 2) {
+          const scale = this.calculateScale(event.nativeEvent.touches);
+          this.scale = scale;
+          this.scaleValue.setValue(scale);
+        }
+      },
+      onPanResponderTerminationRequest: (event, gestureState) => false,
+      onPanResponderRelease: (event, gestureState) => {
+        const scale = clamp(this.scale, this.props.maximumZoomScale, this.props.minimumZoomScale);
+        const verticalLeeway = (scale - 1) * this.size.height / scale / 2;
+        const horizontalLeeway = (scale - 1) * this.size.width / scale / 2;
+        const offset = {
+          x: clamp(this.offset.x, horizontalLeeway, -horizontalLeeway),
+          y: clamp(this.offset.y, verticalLeeway, -verticalLeeway),
+        };
+
+        if (scale !== this.scale) {
+          Animated.timing(this.scaleValue, { toValue: scale, duration: 300 }).start();
+        }
+        if (offset.x !== this.offset.x || offset.y !== this.offset.y) {
+          Animated.spring(this.offsetValue, { toValue: offset, duration: 300 }).start();
+        }
+        this.scale = scale;
+        this.offset = offset;
+      },
+      onShouldBlockNativeResponder: (event, gestureState) => false,
+    });
+  }
+
+  calculateOffset = gestureState => ({
+    x: this.startOffset.x + gestureState.dx / this.scale,
+    y: this.startOffset.y + gestureState.dy / this.scale,
+  });
+
+  calculateScale = touches =>
+    calculateTouchDistance(touches) / this.startDistance * this.startScale;
+
+  handleLayout = event => {
+    const { width, height } = event.nativeEvent.layout;
+    this.size = { width, height };
+  };
+
+  render() {
+    return (
+      <View
+        style={this.props.style || styles.container}
+        onLayout={this.handleLayout}
+        {...this.panResponder.panHandlers}
+      >
+        <Animated.Image
```

Comment:
Should we use `Animated.View` instead so that we can easily support custom Image component ?

Reply:
Good point, I've refactored to be agnostic of underlying component.